### DISCUSSION TITLE
Batch tasks

### DISCUSSION THREAD

**Discussion by geomagilles** [infiniticio/infinitic#264] [Ideas]
Proposal: @Batch Annotation for Task Batching

## Overview

This proposal suggests implementing a batching mechanism for tasks and event listeners, particularly useful for operations like database insertions. The idea is to collect a list of messages before processing them in bulk.

*This could be used only for Pulsar shared-consumers.*

###  Example Usage

```kotlin
// Single-item method
override fun book(cart: CarRentalCart): CarRentalResult {
  TODO() // no implementation needed
}

// Batched method
@Batch(maxMessages = 100, maxSeconds = 1.0)
private fun book(carts: Map<String, CarRentalCart>): Map<String, CarRentalResult> {
  // Implementation for batch booking
}
```

### Key Points

* The batched method should have the same name as the single-item method.
* Parameters for the batched method should be an array of the single-item method's parameter type.
* The return type should be a List of the single-item method's return type.
* The @Batch annotation specifies batching parameters:

  * maxMessage: Maximum number of messages to batch
  * maxDelaySeconds: Maximum delay before processing the batch
 
* If the batch method is present, the single-item method will never be called, and does not need to be actually implemented. 
* The order in the output must *not* changed, as its nth element will be the response of the nth task (for workflows)

## Validation and Behavior

An exception would be raised at runtime:
 * if the batched method doesn't follow the correct signature pattern.
 * it the size of list returned by the backed method is not the same as the input parameters
 
Underlying messages would only be acknowledged when the batched method returns.
The batched method should not be part of the public interface, as it's intended for worker-side use only.

## Configuration Challenges

It's not straightforward to put the batching configuration (maxMessage and maxDelaySeconds) in the worker configuration.
Currently, there are no settings related to methods, only to services.

## Exceptions

If an exception is thrown during the batch processing, then all messages are considered failed

**Comment by Ouriel** [comment]
Hello,
It could be interesting to add to the batch option a way to group event on some properties.
The use-case is the multi-tenancy, in a batch if we have multiple task for different tenant we will need to split the batch in the task to process each tenant separately (different API or database, etc).
The main issue I see is how to read from pulsar with this properties to avoid reading message that will not be consumed fast enough.
It will need to be put as a key in the message and change to key-shared consumer which may not be a good option.

**Comment by geomagilles** [comment]
Following internal discussions, we've made the following important changes to the original proposal:
#### Parameter and Return Type Changes:
The @Batch method now uses a `Map<String, T>` instead of a `List<T>` for both input parameters and return type.
The keys in the map are task IDs, which helps maintain the correct order and association of tasks.
   
#### Optional Batch Key:
An optional additional batch key can now be added through the task metadata.
This feature allows for more flexible grouping of tasks within a batch.

#### Rationale
The change to use Map<String, T> instead of List<T> mitigates the risk of mixing up the order of tasks within a batch.
Task IDs as map keys ensure that each task's input is correctly matched with its output, maintaining consistency and traceability.
The optional batch key in task metadata provides an extra layer of organization and flexibility in batch processing.
